src/fam_struct_wrapper.rs

```diff
@@ -0,0 +1,586 @@
+pub trait FamStruct {
+    /// The type of the zero length array entries
+    type Entry: PartialEq + Copy;
+
+    /// Get the array length
+    ///
+    fn len(&self) -> usize;
+
+    /// Set the array length
+    ///
+    fn set_len(&mut self, len: usize);
+
+    /// Get max array length
+    ///
+    fn max_len() -> usize;
+
+    /// Get the array entries as slice
+    ///
+    fn as_slice(&self) -> &[Self::Entry];
+
+    /// Get the array entries as mut slice
+    ///
+    fn as_mut_slice(&mut self) -> &mut [Self::Entry];
+}
+
+/// A wrapper that helps in treating a `FamStruct` similarly to an actual `Vec`.
+///
+pub struct FamStructWrapper<T: Default + FamStruct> {
+    // this variable holds the FamStruct structure. We use a `Vec<T>` to make the allocation
+    // large enough while still being aligned for `T`. Only the first element of `Vec<T>`
+    // will actually be used as a `T`. The remaining memory in the `Vec<T>` is for `entries`,
+    // which must be contiguous. Since the entries are of type `FamStruct::Entry` we must
+    // be careful to convert the desired capacity of the `FamStructWrapper`
+    // from `FamStruct::Entry` to `T` when reserving or releasing memory.
+    mem_allocator: Vec<T>,
+    // the number of elements of type `FamStruct::Entry` currently in the vec
+    len: usize,
+    // the capacity of the `FamStructWrapper` measured in elements of type `FamStruct::Entry`
+    capacity: usize,
+}
+
+impl<T: Default + FamStruct> FamStructWrapper<T> {
+    /// Get the capacity required by mem_allocator in order to hold
+    /// the provided number of `FamStruct::Entry`
+    ///
+    fn mem_allocator_len(wrapper_len: usize) -> usize {
+        let wrapper_size_in_bytes = size_of::<T>() + wrapper_len * size_of::<T::Entry>();
+        (wrapper_size_in_bytes + size_of::<T>() - 1) / size_of::<T>()
+    }
+
+    /// Get the number of elements of type `FamStruct::Entry` that fit
+    /// in a mem_allocator of provided len
+    ///
+    fn fam_len(mem_allocator_len: usize) -> usize {
+        if mem_allocator_len == 0 {
+            return 0;
+        }
+
+        let array_size_in_bytes = (mem_allocator_len - 1) * size_of::<T>();
+        array_size_in_bytes / size_of::<T::Entry>()
+    }
+
+    /// Constructs a new FamStructWrapper<T> that contains `num_elements`
+    /// default-initialized elements of type `FamStruct::Entry`
+    ///
+    /// # Arguments
+    ///
+    /// * `num_elements` - The number of default-initialized elements of type
+    ///                    `FamStruct::Entry` in the initial `FamStructWrapper`
+    ///
+    pub fn new(num_elements: usize) -> FamStructWrapper<T> {
+        let required_mem_allocator_capacity =
+            FamStructWrapper::<T>::mem_allocator_len(num_elements);
+
+        let mut mem_allocator = Vec::with_capacity(required_mem_allocator_capacity);
+        for _ in 0..required_mem_allocator_capacity {
+            mem_allocator.push(T::default())
+        }
```

Comment:
Should we initialize memory other than first element to zero instead of T::default()? T::default() may generate some non-zero content.